monads/src/HW/Eval.hs

```diff
@@ -18,19 +22,56 @@ type Env v = M.Map v Int -- The environment, mapping variables to their values
 -- The machine state consists of the stack and the variable environment.
 data MachineState v = MachineState
   { getStack :: Stack,
-    getEnv :: Env v
+    getEnv   :: Env v
   }
-  deriving (Show)
+  deriving (Show, Eq)
+
+getStackFromMachineState :: MachineState v -> Stack
+getStackFromMachineState (MachineState getStack _) = getStack
+
+getEnvFromMachineState :: MachineState v -> Env v
+getEnvFromMachineState (MachineState _ getEnv) = getEnv
```

Comment:
You don't need to create these functions: `getStack` and `getEnv` do exactly this.